text/0000-offline-buildpackages.md

```diff
@@ -0,0 +1,479 @@
+
+### Example
+
+```
+description = "builder.toml definition of some builder"
+
+[[buildpacks]]
+  image = "some-buildpack-A-image"
+  version = "1.2.3"
+
+[[buildpacks]]
+ image = "..."
+ version = "..."
+
+[lifecycle]
+  version = "some-version"
+
+[[order]]
+
+  [[order.group]]
+    id = "some-buildpack-A"
+    version = "1.2.3"
+
+[[order]]
+
+  [[order.group]]
+    id = "some-other-buildpack"
+    version = "3.4.5"
+    
+[stack]
+  id = "some-stack"
+  build-image = "some-build-image"
+  run-image = "..."
+  run-image-mirrors = ["..."]
+
+[assets]
+  [[assets.asset-caches]]
+    uri = "some-asset-cache"
+    
+  [[assets.asset-caches]]
+    uri = "some-other-asset-cache"
+    
+  [[assets.exclude-buildpack-assets]]
+    id = "(required)"
+    version = "(required)"
+
+  [[assets.exclude-assets]]
+    sha256 = "some-sha256"
+
+  [[assets.asset-override]]
+    uri = "preferred/asset/location"
+    sha256 = "asset-sha256"
+```
+
+In the above example we can use the `asset-caches` array to specify multiple uris to a number of asset caches. The assets from each of these will be filtered according to the `exclude-buildpack-assets`, `exclude-assets` and `asset-override` fields. assets that pass through these filterings are then added to the resulting builder.
+
+To ease adding assets to a builder we propose adding the `pack create-builder --asset-cache` flag which may specify the uri of an asset cache. This flag may be used multiple times to add multiple asset caches. All assets in the specified caches will then be added to the builder.
+
+Finally there should be the option to include all assets from specified buildpackages using the `--offline` flag (the resulting builders will be very large).
+
+## Using Asset Caches
+
+Asset caches can be added to a build by two mechanisms
+1) specify an asset cache(s) using the `--asset-cache` flag(s) in a `pack build <...>`.
+2) Assets cache layers can be added to a builder image during its creation. These assets will then be available to all builds that use this builder.
+
+
+When asset caches are added to a build/builder we need to verify the following:
+  - The asset is not excluded through a `asset-cache.toml` file passed to `pack build`.
+  - If two assets provide the same `/cnb/assets/<asset-digest>` these two files must have identical contents. Decisions about rewriting these layers to optimize space are left to the platform.
+
+When creating a builder with asset caches:
+  - If any asset layer is a superset of another, only the superset layer is included in the builder.
+  - builders have a `io.buildpacks.buildpack.assets` Label containing entries for every asset included in the builder.
+
+
+### `lifecycle` changes
+Builders should now provide a `CNB_ASSETS` environment variable to the `build` phase`. This provides a standard variable buildpacks may use when looking up assets. The value of `CNB_ASSETS` will default to `/cnb/assets`.
```

Comment:
Won't any alternative to `/cnb/assets` require rewriting the structure of an asset image? What is the usecase for this?

Reply:
This is more for convenience so buildpack authors have a environment variable they can refer to versus. It also gives us a level flexibility if we come up with a more efficient way to provide unarchived assets (which was a part of this RFC but was pulled out)